src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRefreshListener.cs

```diff
@@ -45,9 +47,27 @@ internal class SemanticTokensRefreshListener
 
         private void OnLspWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e) => _workQueue.AddWork();
 
-        // TO-DO: Replace hardcoded string with const once LSP side is merged.
-        public ValueTask SendSemanticTokensNotificationAsync(CancellationToken cancellationToken)
-            => _languageServerNotificationManager.SendNotificationAsync(Methods.WorkspaceSemanticTokensRefreshName, cancellationToken);
+        public async ValueTask SendSemanticTokensNotificationAsync(CancellationToken cancellationToken)
+        {
+            // Currently, we send a refresh notification to the client whenever there are any workspace changes.
+            // However, we also need to send notifications whenever the compilations finish processing since
+            // compilations impact semantic tokens. We put this logic here rather than in the semantic tokens
+            // range handler so that the client still gets initial colorization on file open.
+            var trackedDocuments = _lspWorkspaceManager.GetTrackedLspText();
+            var projects = trackedDocuments.Select(
+                d => _lspWorkspaceManager.GetLspDocument(new TextDocumentIdentifier { Uri = d.Key })?.Project).Distinct();
+
+            foreach (var project in projects)
```

Comment:
This is potentially problematic as we'll be trying to get a lot more compilations because we're now calculating compilations for all changes to all the open projects instead of just the snapshot of the project that is being requested by the semantic tokens range handling. 

I think we definitely only want to trigger the compilation for the project requested by the range handler.

Reply:
> I think we definitely only want to trigger the compilation for the project requested by the range handler.

This is my thinking too, but I'd like to hear the thought process to build all projects in case I'm missing something.